### ISSUE TITLE
Update version of commons-collections dependency in jacoco-maven-plugin

### ISSUE THREAD

**Issue by jlwheat** [jacoco/jacoco#641]
[plugin.xml.txt](https://github.com/jacoco/jacoco/files/1616456/plugin.xml.txt)
The jacoco-maven-plugin uses apache commons collections 3.2, which has been determined to have a security vulnerability.  (Commons-collections-3.2.2 and higher are patched for the vulnerability.)  Our goal is to remove all vulnerable libraries from our local servers, build servers and artifactory, which means we will need to discontinue use of the jacoco maven plugin, unless it is patched.

**The dependency is injected via the \META-INF\maven\plugin.xml file.**

<dependency>
<groupId>commons-collections</groupId>
<artifactId>commons-collections</artifactId>
<type>jar</type>
<version>3.2</version>
</dependency>

JaCoCo version:  jacoco-maven-plugin-0.8.0 and below
Operating system:  Linux
Tool integration: Maven/Ant/API/Other

We need the dependency to point to version 3.2.2 or higher, otherwise, our applications that use the jacoco maven plugin will require the vulnerable library to be present in order to run builds.

===================================================

Here is information outlining the security vulnerability.

WebSphere vulnerability report:
https://www-01.ibm.com/support/docview.wss?uid=swg21970575

Determining if your application uses the Apache Commons InvokerTransformer and vulnerable to CVE-2015-7450
https://www.ibm.com/developerworks/community/blogs/aimsupport/entry/determining_if_your_application_uses_apache_commons_invokertransformer_and_vulnerable_to_cve20157450?lang=en

According to this document, anything prior to Apache Commons  v3.2.2 is vulnerable to the issue.
https://commons.apache.org/proper/commons-collections/javadocs/api-3.2.2/org/apache/commons/collections/functors/InvokerTransformer.html

This is the original publication of the vulnerability:
https://foxglovesecurity.com/2015/11/06/what-do-weblogic-websphere-jboss-jenkins-opennms-and-your-application-have-in-common-this-vulnerability/

**Comment by Godin** [comment]
First of all - this is open source project, if you don't wanna discontinue its usage, then why not submitting patch, helping yourself and others. Or at least be polite and start with such words as "hello".

Second - feel free to replace usage of `jacoco-maven-plugin` by `maven-antrun-plugin` with JaCoCo Ant Tasks that do not depend on commons-collection.

Finally you can update dependencies of Maven plugins without their releases - see https://maven.apache.org/guides/mini/guide-configuring-plugins.html#Using_the_dependencies_Tag :
> Using the `<dependencies>` tag you could configure the dependencies of the Build plugins

More specifically:

```
  <build>
    <pluginManagement>
      <plugins>
        <plugin>
          <groupId>org.jacoco</groupId>
          <artifactId>jacoco-maven-plugin</artifactId>
          <version>0.8.0</version>
          <dependencies>
            <dependency>
              <groupId>commons-collections</groupId>
              <artifactId>commons-collections</artifactId>
              <version>3.2.2</version>
            </dependency>
          </dependencies>
        </plugin>
      </plugins>
    </pluginManagement>
```

**Comment by jlwheat** [comment]
Hello Godin!

Thanks for the quick response and the criticism.  It is awesome and very much appreciated!  

Sorry for my bad manners.  I did not observe this "hello" pattern in any of the other issues I looked at, but I will work on that.  (By the way, you didn't say "hello" either.)

I was planning on submitting a patch.  (Should I have done that before logging the issue?)  I am not as smart as you, so I have not yet figured out how to modify the source, which is in the plugin.xml file that is auto-generated. This file is in the jacoco-maven-plugin-0.8.0.jar that is downloaded, however it is generated by the plugin tool, so it is not available in the source branch in github.  I don't know anything about the plugin tool and am trying to get up to speed on how to do the fix so it can be added to a release.

I will look into the maven-antrun-plugin advice that you gave, and see if that will work for us and our usage of SonarCube on our build servers.  Thanks for that.

Lastly, I work at a big company, so updating the dependencies without a release, might spell trouble, since we have an artifactory that any developer here can upload to, and many of our build jobs reference.   We can set up exclusions in artifactory, forcing builds to break however, without governance, folks could easily circumvent that and acquire a local copy of the vulnerable version.  This could happen anyway, but we sure don't want to encourage it.  This might work if we find that the plugin is not very widely used here.

Thanks!

**Comment by Godin** [comment]
> By the way, you didn't say "hello" either.

That was done on purpose 😉 

> I will look into the maven-antrun-plugin advice that you gave

Please pay attention that in https://github.com/jacoco/jacoco/issues/641#issuecomment-356432384 I also mentioned how your developers can update dependencies of `jacoco-maven-plugin` without waiting release from us, which is simpler than move to Ant Tasks and IMO allows to ban unwanted versions of `commons-collections` in your Artifactory without much pushback from developers.

Or even patch `https://repo.maven.apache.org/maven2/org/jacoco/jacoco-maven-plugin/0.8.0/jacoco-maven-plugin-0.8.0.pom` as
```diff
   <dependencies>
+    <dependency>
+      <groupId>commons-collections</groupId>
+      <artifactId>commons-collections</artifactId>
+      <version>3.2.2</version>
+    </dependency>
     <dependency>
```
and `META-INF/maven/plugin.xml` in `https://repo.maven.apache.org/maven2/org/jacoco/jacoco-maven-plugin/0.8.0/jacoco-maven-plugin-0.8.0.jar` as
```diff
       <type>jar</type>
-      <version>3.2</version>
+      <version>3.2.2</version>
     </dependency>
```
and deploy/use them in your Artifactory.

We released `0.8.0` just recently and need to take care of many other tasks before releasing next version, including other work in this project and in other projects and work in our companies. So even if patch will be committed into master, its immediate release alone is IMO kind of an overkill given that dependency of plugin can be updated without its release as mentioned several times above.

We planned to do next release before/near release of JDK 10, adding support of Java 10, which on its own requires quite some development/testing/etc in various places. Will do our best to also address this issue in it.

Thank you for your understanding.

**Comment by Godin** [comment]
Or BTW even simply substitute `commons-collections:3.2` by `commons-collections:3.2.2` in your Artifactory :wink:

**Comment by marchof** [comment]
This is our current dependency tree for `commons-collections`. So it looks like commons-collections is only referenced by the `maven-reporting-impl` dependency:

<img width="446" alt="image" src="https://user-images.githubusercontent.com/1936262/34758423-fbb615e6-f5d6-11e7-9c8c-77046f077820.png">

**Comment by Godin** [comment]
@marchof yep and that's why I doubt that it can be exploited at all, especially when feature of site rendering not used, which is BTW IMO the case for majority of users.

**Comment by jlwheat** [comment]
Hi all!

Sorry if this issue I brought up was more annoying than informative.  That wasn't my intention.

After working on this further with my code base, I found a few more plugins that were dependent on commons-collections-3.2.1 and older.  That made things a bit more complex to fix on my end, again with the goal of getting the older jars off of our company's sandboxes, build servers, and artifactory.  (And, of course, our production servers.)

I ended up using transitive dependency overrides in maven.  (Per Godin's suggestion.)  This is simple and effective, once you are able to determine which libraries and plugins are pulling in the dependencies. (I know.  Duh, right?)  Here is a list of the ones I found.  I would imagine there is a slough of others.

OpenEJB -  fixed in 4.7.5
OpenJPA - fixed in 2.4.1 
Openejb-junit - fixed in 4.7.5
Jacoco-maven-plugin - no patch as of 0.7.9
Maven-site-plugin - no patch as of 3.7
Maven-javadoc-plugin - no patch as of 3.0.0
Maven-plugin-plugin - no patch as of 3.5

And commons-collections itself was patched in version 3.2.2.  

I found I was able to override the dependencies on the unpatched plugins fairly easily.  (And, of course,
I switched to the patched versions for the libraries.)  I was able to do this in our templates and custom framework, which fixes the applications that reference those during the build.

        <plugin>
                <groupId>org.jacoco</groupId>
                <artifactId>jacoco-maven-plugin</artifactId>
                <version>0.7.5.201505241946</version>
                <dependencies>
                     <dependency>
			<groupId>commons-collections</groupId>
			<artifactId>commons-collections</artifactId>
			<version>${commons.collections.version}</version>
		     </dependency>
		</dependencies>
                <executions>
                    <execution>
                        <id>jacoco-initialize</id>
                        <goals>
                            <goal>prepare-agent</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>

![apachecommonssecurityvulnerability](https://user-images.githubusercontent.com/22482413/35128704-d7c6ba68-fc6c-11e7-96cc-3dec79376632.png)

I did this for all the plugins that have the dependency.  Then set the commons.collections.version property to 3.2.2 in the parent poms.  Once this is finalized, we can safely remove the older versions of commons-collections from the servers, including artifactory.  Then, if we encounter build issues in our applications in the future, we can point them to the new version of our custom framework 
which has the overrides.

Thanks Godin and marchof!
